File: MoreHighlight/src/main/java/xyz/wingio/plugins/morehighlight/ColorRule.java
package xyz.wingio.plugins.morehighlight;

import android.content.Context;
import android.graphics.Color;

import com.aliucord.PluginManager;
import com.discord.simpleast.core.parser.ParseSpec;
import com.discord.simpleast.core.parser.Parser;
import com.discord.simpleast.core.node.Node;
import com.discord.simpleast.core.parser.Rule;
import com.discord.utilities.textprocessing.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class ColorRule extends Rule<MessageRenderContext, ColorNode<MessageRenderContext>,MessageParseState> {
    public ColorRule() {
        super(Pattern.compile("^#[0-9a-fA-F]{6,8}"));
    }

    @Override
    public ParseSpec<MessageRenderContext, MessageParseState> parse(Matcher matcher, Parser<MessageRenderContext, ? super ColorNode<MessageRenderContext>, MessageParseState> parser, MessageParseState s) {
        try {
            ColorNode textNode = new ColorNode(matcher.group(0), Color.parseColor(matcher.group(0)));
            return new ParseSpec<>(textNode, s);
        } catch (Exception e) {
            return new ParseSpec<>(new TextNode(matcher.group(0)), s);
        }
    }
}

File: MoreHighlight/src/main/java/xyz/wingio/plugins/morehighlight/rule/ColorRule.kt
package xyz.wingio.plugins.morehighlight.rule

import android.graphics.Color
import com.discord.simpleast.core.parser.ParseSpec
import com.discord.simpleast.core.parser.Parser
import com.discord.simpleast.core.parser.Rule
import com.discord.utilities.textprocessing.MessageParseState
import com.discord.utilities.textprocessing.MessageRenderContext
import xyz.wingio.plugins.morehighlight.node.ColorNode
import xyz.wingio.plugins.morehighlight.node.TextNode
import java.util.regex.Matcher
import java.util.regex.Pattern

class ColorRule: Rule<MessageRenderContext, ColorNode<MessageRenderContext?>, MessageParseState>(
    Pattern.compile("^#[0-9a-fA-F]{6,8}")
) {
    
    override fun parse(
        matcher: Matcher,
        parser: Parser<MessageRenderContext, in ColorNode<MessageRenderContext?>?, MessageParseState>,
        s: MessageParseState
    ): ParseSpec<MessageRenderContext, MessageParseState> {
        return try {
            ParseSpec(ColorNode(matcher.group(0)!!, Color.parseColor(matcher.group(0))), s)
        } catch (_: Exception) {
            ParseSpec(TextNode(matcher.group(0)!!), s)
        }
    }

}